code/game/objects/items/boombox.dm

```diff
@@ -0,0 +1,245 @@
+//A lot of code taken from jukebox
+//This is not a child object because it's an item
+
+/obj/item/media/boombox
+	name = "BL4ST4-TUN3 3000"
+	desc = "A snazzy boombox, hefty enough to be carried around your shoulder."
+	icon = 'icons/obj/boombox.dmi'
+	icon_state = "boombox"
+	item_state = "boombox"
+
+	var/loop_mode = JUKEMODE_PLAY_ONCE			// Behavior when finished playing a song
+	var/max_queue_len = 3						// How many songs are we allowed to queue up?
+	var/datum/track/current_track				// Currently playing song
+	var/list/datum/track/queue = list()			// Queued songs
+	var/list/datum/track/tracks = list()		// Available tracks
+
+	var/list/current_listeners = list() //People listening to boombox right now
+
+	var/ticks_to_update = 5
+
+	var/obj/item/weapon/music_tape/my_tape //boombox tape
+
+	var/sanity_value = 2
+
+/obj/item/media/boombox/New()
+	AddComponent(/datum/component/atom_sanity, 0, "")
+	START_PROCESSING(SSobj, src)
+
+/obj/item/media/boombox/Destroy()
+	STOP_PROCESSING(SSobj, src)
+
+/obj/item/media/boombox/Process()
+	if(ticks_to_update)
+		ticks_to_update--
+	else
+		update_music()
+		ticks_to_update = 3
+	if(!playing)
+		return
+	// If the current track isn't finished playing, let it keep going
+	if(current_track && world.time < media_start_time + current_track.duration)
+		return
+	// Otherwise time to pick a new one!
+	if(queue.len > 0)
+		current_track = queue[1]
+		queue.Cut(1, 2)  // Remove the item we just took off the list
+	else
+		// Oh... nothing in queue? Well then pick next according to our rules
+		switch(loop_mode)
+			if(JUKEMODE_NEXT)
+				var/curTrackIndex = max(1, tracks.Find(current_track))
+				var/newTrackIndex = (curTrackIndex % tracks.len) + 1  // Loop back around if past end
+				current_track = tracks[newTrackIndex]
+			if(JUKEMODE_RANDOM)
+				var/previous_track = current_track
+				do
```

Comment:
What does `do` do?

Reply:
17.4 do while loop
The do while loop is similar to the while loop, except the condition is tested at the end of an iteration rather than the beginning. The effect this has is to guarantee that the body of the loop is executed at least once. In certain situations, that is just what one needs.